src/operator/subgraph/dnnl/dnnl_transformer_qk_common.h

```diff
@@ -0,0 +1,230 @@
+#include "operator/tensor/matrix_op-inl.h"
+#include "operator/subgraph/common.h"
+#include "dnnl_common.h"
+#include "dnnl_subgraph_base-inl.h"
+#include "dnnl_transformer-inl.h"
+
+namespace mxnet {
+namespace op {
+namespace qk_common {
+
+enum SelectStatusTransformerQK {
+  kFail = 0,
+  kStart,
+  kFirstSwapAx,
+  kSecondSwapAx,
+  kFirstReshape,
+  kSecondReshape,
+  kSuccess
+};
+
+// /*
+// kStart ---> kFirstSwapAx ---> kSecondSwapAx ---> kFirstReshape ---> kSecondReshape ---> kSuccess
+// OR
+// kStart ---> kFirstSwapAx ---> kSecondSwapAx ---> kFirstReshape ---> kSuccess
+// each status except kStart is connected with kFail
+// */
+
+inline bool CheckSwapAxisConditionsQK(const BiDirectedNode& input_node) {
+  if (input_node.outputs.size() != 1)
+    return false;
+  return CheckSwapAxisConditions(*input_node.node);
+}
+
+inline bool CheckReshapeConditionsQK(const BiDirectedNode& input_node, const index_t out_index) {
+  if (input_node.outputs.size() != 1)
+    return false;
+  return CheckReshapeConditions(*input_node.node, out_index);
+}
+
+inline bool CheckSplitConditions(const std::vector<const BiDirectedNode*>& matched_list,
+                                 const BiDirectedNode& node) {
+  const SplitParam& param = dmlc::get<SplitParam>(node.node->attrs.parsed);
+
+  if (param.axis != -1 || param.sections != 3 || param.squeeze_axis)
+    return false;
+
+  const auto first_reshape  = (*(matched_list.end() - 2))->node;
+  const auto second_reshape = (*(matched_list.end() - 1))->node;
+  if (first_reshape->op() != Op::Get("_npx_reshape") ||
+      second_reshape->op() != Op::Get("_npx_reshape")) {
+    return false;
+  }
+  // 3 sections - ensure that every output is used only once
+  if (node.outputs.size() == 3 && node.outputs.count(first_reshape) &&
+      node.outputs.count(second_reshape)) {
+    return true;
+  }
+
+  return false;
+}
+
+inline bool Select(SelectStatusTransformerQK* status,
+                   std::vector<const BiDirectedNode*>* matched_list,
+                   const BiDirectedNode& seed_node,
+                   const std::shared_ptr<NodeAttr>& node_attr) {
+  if (seed_node.node->op() == Op::Get("batch_dot")) {
+    *status = kStart;
+    matched_list->clear();
+    matched_list->push_back(&seed_node);
+    return true;
+  }
+  return false;
+}
+
+template <bool with_split>
+bool SelectInput(SelectStatusTransformerQK* status,
+                 std::vector<const BiDirectedNode*>* matched_list,
+                 const BiDirectedNode& n,
+                 const BiDirectedNode& input_node) {
+  if (*status == kFail || *status == kSuccess || input_node.node->is_variable())
+    return false;
+  const auto& raw_input_node = *input_node.node;
+  switch (*status) {
+    case kStart:
+      if (raw_input_node.op() == Op::Get("SwapAxis")) {
+        if (CheckSwapAxisConditionsQK(input_node)) {
+          *status = kFirstSwapAx;
+          matched_list->push_back(&input_node);
+        }
+        return true;
```

Comment:
Shouldn't this return true be in 'if' statement below matched_list->push_back(&input_node); ?

Reply:
done